packages/babel-core/src/config/option-manager.js

```diff
@@ -74,340 +61,398 @@ const optionNames = new Set([
+  }
 
-    if (type === "preset") {
-      if (opts.only !== undefined) throw new Error(`${alias}.only is not supported in a preset`);
-      if (opts.ignore !== undefined) throw new Error(`${alias}.ignore is not supported in a preset`);
-      if (opts.extends !== undefined) throw new Error(`${alias}.extends is not supported in a preset`);
-      if (opts.env !== undefined) throw new Error(`${alias}.env is not supported in a preset`);
-    }
+  if (!item || typeof item !== "object") {
+    throw new Error("Plugin/Preset did not return an object.");
+  }
 
-    if (opts.sourceMap !== undefined) {
-      if (opts.sourceMaps !== undefined) {
-        throw new Error(`Both ${alias}.sourceMap and .sourceMaps have been set`);
-      }
+  return { value: item, descriptor };
+}
 
-      opts.sourceMaps = opts.sourceMap;
-      delete opts.sourceMap;
-    }
+/**
+ * Instantiate a plugin for the given descriptor, returning the plugin/options pair.
+ */
+const PLUGIN_CACHE = new WeakMap();
+function loadPluginDescriptor(descriptor) {
+  if (descriptor.value instanceof Plugin) return [ descriptor.value, descriptor.options ];
+
+  let result = PLUGIN_CACHE.get(descriptor.value);
+  if (!result) {
+    result = instantiatePlugin(loadDescriptor(descriptor, true /* skipOptions */));
+    PLUGIN_CACHE.set(descriptor.value, result);
+  }
 
-    for (const key in opts) {
-      // check for an unknown option
-      if (!optionNames.has(key)) {
-        if (removed[key]) {
-          throw new ReferenceError(`Using removed Babel 5 option: ${alias}.${key} - ${removed[key].message}`);
-        } else {
-          // eslint-disable-next-line max-len
-          const unknownOptErr = `Unknown option: ${alias}.${key}. Check out http://babeljs.io/docs/usage/options/ for more information about options.`;
-
-          throw new ReferenceError(unknownOptErr);
-        }
-      }
-    }
+  return [ result, descriptor.options];
+}
 
-    if (opts.parserOpts && typeof opts.parserOpts.parser === "string") {
-      opts.parserOpts.parser = loadParser(opts.parserOpts.parser, dirname).parser;
+function instantiatePlugin({ value: pluginObject, descriptor }) {
+  Object.keys(pluginObject).forEach((key) => {
+    if (!ALLOWED_PLUGIN_KEYS.has(key)) {
+      throw new Error(messages.get("pluginInvalidProperty", descriptor.alias, key));
     }
+  });
+  if (pluginObject.visitor && (pluginObject.visitor.enter || pluginObject.visitor.exit)) {
+    throw new Error("Plugins aren't allowed to specify catch-all enter/exit handlers. " +
+      "Please target individual nodes.");
+  }
 
-    if (opts.generatorOpts && typeof opts.generatorOpts.generator === "string") {
-      opts.generatorOpts.generator = loadGenerator(opts.generatorOpts.generator, dirname).generator;
-    }
+  const plugin = Object.assign({}, pluginObject, {
+    visitor: clone(pluginObject.visitor || {}),
+  });
+
+  traverse.explode(plugin.visitor);
+
+  let inheritsDescriptor;
+  let inherits;
+  if (plugin.inherits) {
```

Comment:
I had the discussion in a PR last time, could we support multiple inheritance?

```js
{
  inherits: require('foo')
}
```

vs

```js
{
  inherits: [
    require('foo'),
    require('bar'),
  ]
}
```